File: Sources/StitchPilotCore/Model/Geometry.swift
import Foundation

/// A point in the design's physical coordinate space, in millimeters.
/// Millimeters (not pixels, not machine units) are the canonical unit
/// throughout the internal model — see ARCHITECTURE.md "Units" section.
public struct Point2D: Codable, Hashable, Sendable {
    public var x: Double
    public var y: Double

    public init(_ x: Double, _ y: Double) {
        self.x = x
        self.y = y
    }

    public static let zero = Point2D(0, 0)

    public func distance(to other: Point2D) -> Double {
        (self - other).length
    }

    public var length: Double { (x * x + y * y).squareRoot() }

    public static func + (a: Point2D, b: Point2D) -> Point2D { Point2D(a.x + b.x, a.y + b.y) }
    public static func - (a: Point2D, b: Point2D) -> Point2D { Point2D(a.x - b.x, a.y - b.y) }
    public static func * (a: Point2D, s: Double) -> Point2D { Point2D(a.x * s, a.y * s) }
}

/// A single open or closed polyline, in millimeters, in object-local space.
/// Curves from source vector artwork are flattened to polylines at import
/// time (with a fine tolerance) rather than rasterized — geometry stays
/// vector-precision, just represented piecewise-linearly so every downstream
/// stitch algorithm (resampling, offsetting, satin-rail generation) can work
/// with plain line segments instead of re-deriving curve math everywhere.
public struct SubPath: Codable, Hashable, Sendable {
    public var points: [Point2D]
    public var closed: Bool

    public init(points: [Point2D], closed: Bool) {
        self.points = points
        self.closed = closed
    }

    public var length: Double {
        guard points.count > 1 else { return 0 }
        var total = 0.0
        for i in 1..<points.count { total += points[i - 1].distance(to: points[i]) }
        if closed, let first = points.first, let last = points.last {
            total += last.distance(to: first)
        }
        return total
    }

    public var boundingBox: BoundingBox {
        BoundingBox(points: points)
    }
}

/// The geometric shape of one embroidery object: one or more sub-paths
/// (e.g. a letter "O" is an outer subpath plus an inner hole subpath).
public struct VectorShape: Codable, Hashable, Sendable {
    public var subPaths: [SubPath]

    public init(subPaths: [SubPath]) {
        self.subPaths = subPaths
    }

    public var boundingBox: BoundingBox {
        var box = BoundingBox.empty
        for sp in subPaths { box = box.union(sp.boundingBox) }
        return box
    }
}

public struct BoundingBox: Codable, Hashable, Sendable {
    public var minX, minY, maxX, maxY: Double

    public static let empty = BoundingBox(minX: .infinity, minY: .infinity, maxX: -.infinity, maxY: -.infinity)

    public init(minX: Double, minY: Double, maxX: Double, maxY: Double) {
        self.minX = minX; self.minY = minY; self.maxX = maxX; self.maxY = maxY
    }

    public init(points: [Point2D]) {
        self = points.reduce(BoundingBox.empty) { box, p in
            BoundingBox(minX: min(box.minX, p.x), minY: min(box.minY, p.y),
                        maxX: max(box.maxX, p.x), maxY: max(box.maxY, p.y))
        }
    }

    public var width: Double { maxX - minX }
    public var height: Double { maxY - minY }
    public var isEmpty: Bool { minX > maxX || minY > maxY }

    public func union(_ other: BoundingBox) -> BoundingBox {
        if isEmpty { return other }
        if other.isEmpty { return self }
        return BoundingBox(minX: min(minX, other.minX), minY: min(minY, other.minY),
                            maxX: max(maxX, other.maxX), maxY: max(maxY, other.maxY))
    }
}

File: web/src/geometry.ts
import type { BoundingBox, EmbroideryObject, Point2D, VectorShape } from "./types";

export const EMPTY_BOX: BoundingBox = { minX: Infinity, minY: Infinity, maxX: -Infinity, maxY: -Infinity };

export function shapeBounds(shape: VectorShape): BoundingBox {
  let b = { ...EMPTY_BOX };
  for (const sp of shape.subPaths) for (const p of sp.points) {
    if (p.x < b.minX) b.minX = p.x; if (p.x > b.maxX) b.maxX = p.x;
    if (p.y < b.minY) b.minY = p.y; if (p.y > b.maxY) b.maxY = p.y;
  }
  return b;
}

export function unionBounds(a: BoundingBox, b: BoundingBox): BoundingBox {
  return { minX: Math.min(a.minX, b.minX), minY: Math.min(a.minY, b.minY), maxX: Math.max(a.maxX, b.maxX), maxY: Math.max(a.maxY, b.maxY) };
}

export const boxIsEmpty = (b: BoundingBox) => b.minX > b.maxX || b.minY > b.maxY;
export const boxesIntersect = (a: BoundingBox, b: BoundingBox) => a.minX <= b.maxX && b.minX <= a.maxX && a.minY <= b.maxY && b.minY <= a.maxY;

/** Even-odd rule across every sub-path, so a letter's counter reads as a hole. */
export function shapeContains(shape: VectorShape, p: Point2D): boolean {
  let inside = false;
  for (const sp of shape.subPaths) {
    const pts = sp.points;
    for (let i = 0, j = pts.length - 1; i < pts.length; j = i++) {
      const a = pts[i], b = pts[j];
      if ((a.y > p.y) !== (b.y > p.y) && p.x < ((b.x - a.x) * (p.y - a.y)) / (b.y - a.y) + a.x) inside = !inside;
    }
  }
  return inside;
}

/** Topmost object under the point (later objects sew on top). */
export function objectAt(objects: EmbroideryObject[], p: Point2D): EmbroideryObject | null {
  for (let i = objects.length - 1; i >= 0; i--) {
    const b = shapeBounds(objects[i].shape);
    if (p.x < b.minX || p.x > b.maxX || p.y < b.minY || p.y > b.maxY) continue;
    if (shapeContains(objects[i].shape, p)) return objects[i];
  }
  return null;
}

export function transformShape(shape: VectorShape, fn: (p: Point2D) => Point2D): VectorShape {
  return { subPaths: shape.subPaths.map((sp) => ({ ...sp, points: sp.points.map(fn) })) };
}

export function selectionBounds(objects: EmbroideryObject[], ids: Set<string>): BoundingBox {
  let b = { ...EMPTY_BOX };
  for (const o of objects) if (ids.has(o.id)) b = unionBounds(b, shapeBounds(o.shape));
  return b;
}
